crates/ruff/src/jupyter/index.rs

```diff
@@ -0,0 +1,83 @@
+use std::iter;
+use std::sync::Arc;
+
+use crate::jupyter::{Cell, SourceValue};
+
+/// Jupyter Notebook indexing table
+///
+/// When we lint a jupyter notebook, we have to translate the row/column based on
+/// [`ruff_text_size::TextSize`] to jupyter notebook cell/row/column.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct JupyterIndex {
+    inner: Arc<JupyterIndexInner>,
+}
+
+impl JupyterIndex {
+    pub fn get_cell(&self, row: usize) -> u32 {
+        self.inner.row_to_cell[row]
+    }
+
+    pub fn get_row_in_cell(&self, row: usize) -> u32 {
+        self.inner.row_to_row_in_cell[row]
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct JupyterIndexInner {
+    /// Enter a row (1-based), get back the cell (1-based)
+    row_to_cell: Vec<u32>,
+    /// Enter a row (1-based), get back the cell (1-based)
+    row_to_row_in_cell: Vec<u32>,
+}
+
+/// Builder for [`JupyterIndex`].
+pub struct JupyterIndexBuilder {
+    row_to_cell: Vec<u32>,
+    row_to_row_in_cell: Vec<u32>,
+}
+
+impl Default for JupyterIndexBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl JupyterIndexBuilder {
```

Comment:
Could we do without the separate builder here?

Reply:
The reason to have a builder is to have both the index and cell offsets built in a single loop with each individual cell.

Cell offsets are built from the source code only for the first time. They cannot be refreshed using the transformed source code but require to process each individual edit to update the offsets.

When a single auto-fix iteration ends, the index is refreshed from the transformed source code, source code is updated in each cell and cell offsets are updated from the edits (order is important).